src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs

```diff
@@ -1396,35 +1400,63 @@ private async Task WriteStreamPropertiesAsync(SelectExpandNode selectExpandNode,
         {
             ISet<IEdmNavigationProperty> navigationProperties = selectExpandNode.SelectedNavigationProperties;
 
-            if (navigationProperties != null)
+            if (navigationProperties == null)
             {
-                IEnumerable<string> changedProperties = null;
-                if (null != resourceContext.EdmObject && resourceContext.EdmObject is IDelta changedObject)
+                yield break;
+            }
+
+            if (resourceContext.EdmObject is IDelta changedObject)
+            {
+                IEnumerable<string> changedProperties = changedObject.GetChangedPropertyNames();
+
+                foreach (IEdmNavigationProperty navigationProperty in navigationProperties)
                 {
-                    changedProperties = changedObject.GetChangedPropertyNames();
+                    if (changedProperties == null || changedProperties.Contains(navigationProperty.Name))
+                    {
+                        yield return new KeyValuePair<IEdmNavigationProperty, Type>(navigationProperty, typeof(IEdmChangedObject));
+                    }
+                }
+            }
+            else if (resourceContext.ResourceInstance is IDelta deltaObject)
+            {
+                IEnumerable<string> changedProperties = deltaObject.GetChangedPropertyNames();
+                dynamic delta = deltaObject;
+
+                foreach (IEdmNavigationProperty navigationProperty in navigationProperties)
+                {
+                    object obj = null;
 
-                    foreach (IEdmNavigationProperty navigationProperty in navigationProperties)
+                    if (changedProperties == null || changedProperties.Contains(navigationProperty.Name) && delta.DeltaNestedResources.TryGetValue(navigationProperty.Name, out obj))
                     {
-                        if (changedProperties == null || changedProperties.Contains(navigationProperty.Name))
-                        {
-                            yield return new KeyValuePair<IEdmNavigationProperty, Type>(navigationProperty, typeof(IEdmChangedObject));
-                        }
+                        yield return new KeyValuePair<IEdmNavigationProperty, Type>(navigationProperty, obj.GetType());
```

Comment:
Could you clarify the change from `typeof(IEdmChangeObject)` to `obj.GetType()`

Reply:
The type didn't change. I refactored the method.
In the `if` code block, we have `typeof(IEdmChangeObject)`
In the `else if` code block, we have `obj.GetType()`